### PR TITLE
Add default .scala-steward.conf for Scala Steward itself

### PR DESCRIPTION
**Pull request by eugeniyk**
In some scenarios there is a need to limit dependencies you want to update for all repos. 

Currently you'd need to update repo configuration (`.scala-steward.conf`) for every project you are working on.
If the repo doesn't have configuration (`maybeRepoConfig`) it will fallback to empty `RepoConfig.default`.

This PR is adding ability to override fallback to `RepoConfig.default` by reading `default.scala-steward.conf` within main scala steward repo. It gives you all the features as described in [repo-specific-configuration.md](https://github.com/fthomas/scala-steward/blob/master/docs/repo-specific-configuration.md)

Specifically, you can limit dependencies by providing `allow` settings (which is helpful if you have 100+ repos and you want to propagate specific dependencies, only)

### Changes
- Added `--default-repo-config` setting (default to `default.scala-steward.conf`)
- Chain of configuration: check repoConf (`$repo/.scala-steward.conf`) => fallback to DefaultRepoConf (`root/default.scala-steward.conf`) => fallback to empty configuration
- Added extra tests

### CODE REVIEW

**Comment by mzuehlke** [comment]
I really like this idea and it will solve #1470.

Have you had a look how difficult it is to merge local configuration with global one ? 
So that in the case of an available local configuration it would override the global one ?

The only minor comment I have after going through the PR is that the default global scala-teward.conf you are referencing in the docs and examples start with a dot. Which would hide i. I would prefer "default.scala-teward.conf" or "global.scala-teward.conf"

**Comment by eugeniyk** [comment]
@mzuehlke thanks for feedback

I think it could be nice continuation of this feature (I see some complexity with "mergeability" for some settings, see below)

Seems so far we have `allow` / `ignore` / `pin` options to be merged, rest should be just repo (local) over global:

1. **Allow** - I'd vote for **superset** since each project knows which the only repos it want to get, and each steward knows which dependencies it wants to operate.
In case of different combinations of `artifact`, `artifact / module`, `artifact / module / revision` we should pick the most restrictive one.
If one of Local / global settings are empty, just pick another one.

2. **Ignore** - union
3. **pin** - union with Local preference in terms of revision

At the same time, let me rename config to be `global.scala-steward.conf`, makes sense

**Comment by mzuehlke** [comment]
Thanks for the detailed investigation. Sound feasible.
But I'm totally fine with doing this in a follow up PR.

**Comment by eugeniyk** [comment]
@mzuehlke updated

**Comment by mzuehlke** [comment]
Thanks @eugeniyk 
I have no possibility to approve or merge.
This is up to @fthomas

**Comment by MPV** [comment]
Author of #1470 here. Looks like this will solve that issue. Just wanted to reach out to say thank you for this PR/contribution. ❤️

File: scripts/run.sh [hunk 1]
```diff
@@ -23,6 +23,8 @@ LOGIN="scala-steward"
 
 COMMON_ARGS=(
   --workspace "$STEWARD_DIR/workspace"
+  --repos-file "$STEWARD_DIR/repos.md"
+  --repos-default-conf "$STEWARD_DIR/default.scala-steward.conf"
```
Comment (fthomas) [inline scripts/run.sh:27]: Could you please revert these two changes. I'm using this script to run my public Scala Steward instance. The `--repos-file` parameter is set later in this file and my instance will not use a default repo config.

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala [hunk 2]
```diff
@@ -16,34 +16,49 @@
...
-      val maybeRepoConfig = OptionT(fileAlg.readFile(configFile)).map(parseRepoConfig).flatMapF {
-        case Right(config)  => logger.info(s"Parsed $config").as(config.some)
-        case Left(errorMsg) => logger.info(errorMsg).as(none[RepoConfig])
-      }
-      maybeRepoConfig.value
+    workspaceAlg
+      .repoDir(repo)
+      .flatMap(projectRootDir => readConfiguration(projectRootDir / repoConfigBasename))
+
+  private def readConfiguration(configFile: File): F[Option[RepoConfig]] = {
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:55]: I would prefer if this is called `readFromFile` or `readRepoConfigFromFile`.

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfig.scala [hunk 3]
```diff
@@ -31,7 +31,7 @@ final case class RepoConfig(
 }
 
 object RepoConfig {
-  val default: RepoConfig = RepoConfig()
+  private[core] val default: RepoConfig = RepoConfig()
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfig.scala:34]: What is the purpose of this change? Is it because we now have two defaults, `RepConfig.default` and `RepoConfigAlg.defaultRepoConfig`, and that is confusing? Would it be less confusing if this is renamed to `RepoConfig.empty`?

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala [hunk 4]
```diff
@@ -16,34 +16,49 @@
...
 ) {
   def readRepoConfigOrDefault(repo: Repo): F[RepoConfig] =
-    readRepoConfig(repo).map(_.getOrElse(RepoConfig.default))
+    readRepoConfig(repo).flatMap { config =>
+      config.map(F.pure).getOrElse(defaultRepoConfig())
+    }
+
+  /**
+    * Default configuration will try to read file specified in config.reposDefaultConfig first; if not found - fallback to empty configuration.
+    * Note it's not lazy since we want to get config updates if you decide to change config in the middle of the process
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:45]: Could you please break lines after 80 characters?

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala [hunk 5]
```diff
@@ -16,34 +16,49 @@
...
-    workspaceAlg.repoDir(repo).flatMap { dir =>
-      val configFile = dir / repoConfigBasename
-      val maybeRepoConfig = OptionT(fileAlg.readFile(configFile)).map(parseRepoConfig).flatMapF {
-        case Right(config)  => logger.info(s"Parsed $config").as(config.some)
-        case Left(errorMsg) => logger.info(errorMsg).as(none[RepoConfig])
-      }
-      maybeRepoConfig.value
+    workspaceAlg
+      .repoDir(repo)
+      .flatMap(projectRootDir => readConfiguration(projectRootDir / repoConfigBasename))
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:53]: Could you rename `projectRootDir` to `repoDir` or just `dir`? I think these names are almost exclusively used in the code base.

File: modules/core/src/main/scala/org/scalasteward/core/application/Cli.scala [hunk 6]
```diff
@@ -37,6 +37,7 @@ object Cli {
   final case class Args(
       workspace: String,
       reposFile: String,
+      reposDefaultConf: String = "default.scala-steward.conf",
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/application/Cli.scala:40]: I would prefer calling this `defaultRepoConf`.

File: modules/core/src/main/scala/org/scalasteward/core/application/Config.scala [hunk 7]
```diff
@@ -49,6 +54,7 @@ import scala.sys.process.Process
 final case class Config(
     workspace: File,
     reposFile: File,
+    reposDefaultConfigFile: File,
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/application/Config.scala:57]: Likewise, I would prefer calling this `defaultRepoConfigFile`.

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala [hunk 8]
```diff
@@ -16,34 +16,49 @@
...
-    readRepoConfig(repo).map(_.getOrElse(RepoConfig.default))
+    readRepoConfig(repo).flatMap { config =>
+      config.map(F.pure).getOrElse(defaultRepoConfig())
+    }
+
+  /**
+    * Default configuration will try to read file specified in config.reposDefaultConfig first; if not found - fallback to empty configuration.
+    * Note it's not lazy since we want to get config updates if you decide to change config in the middle of the process
+    */
+  def defaultRepoConfig(): F[RepoConfig] =
```
[the remark this answers is not in the document]
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:47]: > Note it's not lazy since we want to get config updates if you decide to change config in the middle of the process

I can move it to be val - just thinking if you have 100+ repos and you found an issue or smth, it will give you hot reload

wdyt?

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfig.scala (hunk 3 above)
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfig.scala:34]: Just changed visibility, actually I want it to be used only by `RepoConfigAlg`, so I'll make it `private[repoconfig]`
`RepoConfig.empty` makes more sense for me, will rename

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala (hunk 4 above)
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:45]: No problem
Do you want to change https://github.com/fthomas/scala-steward/blob/master/.scalafmt.conf#L5 to be 80?

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (fthomas) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:48]: Since the type is still `F[RepoConfig]` it will not be eagerly evaluated if this is changed to a `val`.

[the remark this answers is not in the document]
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:48]: you mean even if I leave it as is it won't be eager even if it's `def`?

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala (hunk 4 above)
Comment (fthomas) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:46]: I thought this setting has no effect on Scaladocs?
We shouldn't change that setting in this PR.

Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:46]: it has, otherwise I'd have it formatted by `fmt`

File: modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala:48]: ok set it to val for now

[the remark this answers is not in the document]
Comment (fthomas) [reply modules/core/src/main/scala/org/scalasteward/core/repoconfig/RepoConfigAlg.scala]: I mean that it doesn't semantically matter if this is a `def` or `val` because the effects of that statement are not executed when it is evaluated but when the `F` is run.

File: modules/core/src/main/scala/org/scalasteward/core/update/PruningAlg.scala [hunk 9]
```diff
@@ -64,10 +70,12 @@ final class PruningAlg[F[_]](implicit
       repoCache: RepoCache,
       dependencies: List[Scope.Dependency]
   ): F[(Boolean, List[Update.Single])] = {
-    val repoConfig = repoCache.maybeRepoConfig.getOrElse(RepoConfig.default)
+
     val depsWithoutResolvers = dependencies.map(_.value).distinct
     for {
       _ <- logger.info(s"Find updates for ${repo.show}")
+      defaultConfig <- repoConfigAlg.defaultRepoConfig
+      repoConfig = repoCache.maybeRepoConfig.getOrElse(defaultConfig)
```
Comment (fthomas) [inline modules/core/src/main/scala/org/scalasteward/core/update/PruningAlg.scala:78]: One last thing I noticed now: If `repoCache.maybeRepoConfig` is a `Some` there is no need to load the default repo config. We could change these two lines to:
```scala
repoConfig <- OptionT.fromOption[F](repoCache.maybeRepoConfig)
  .getOrElseF(repoConfigAlg.defaultRepoConfig)
```

Comment (eugeniyk) [reply modules/core/src/main/scala/org/scalasteward/core/update/PruningAlg.scala:78]: actually, in this case we should also merge configs
let me change

**Comment by fthomas** [comment]
@eugeniyk Btw, you can fix the formatting by running `sbt fmt`.

**Comment by eugeniyk** [comment]
I actually fix formatted version ;)

**Comment by eugeniyk** [comment]
@fthomas you have quite advanced defence here: wart, fmt, codacy ;)

**Comment by anilkumarmyla** [comment]
Wow, this PR also solves our use-case in https://github.com/fthomas/scala-steward/issues/1411 , this is awesome. Thanks @eugeniyk

File: build.sbt [hunk 10]
```diff
@@ -236,6 +236,7 @@ runSteward := Def.taskDyn {
   val args = Seq(
     Seq("--workspace", s"$projectDir/workspace"),
     Seq("--repos-file", s"$projectDir/repos.md"),
+    Seq("--default-repo-config", s"$projectDir/default.scala-steward.conf"),
```
Comment (anilkumarmyla) [inline build.sbt:239]: should this be `--default-repo-conf` sans `ig` ?

Comment (fthomas) [reply build.sbt:239]: Yes, good catch. I'll update it.
